src/main/flight/pid.c

```diff
@@ -1416,6 +1431,10 @@ void FAST_CODE pidController(const pidProfile_t *pidProfile, timeUs_t currentTim
             // no transition if feedForwardTransition == 0
             float transition = feedForwardTransition > 0 ? MIN(1.f, getRcDeflectionAbs(axis) * feedForwardTransition) : 1;
             pidData[axis].F = feedforwardGain * transition * pidSetpointDelta * pidFrequency;
+            if (ff_boost_factor && (axis <= FD_PITCH)) {
+                const float F_hpf = pidData[axis].F - pt1FilterApply(&ff_lpf[axis], pidData[axis].F);
+                pidData[axis].F += F_hpf * ff_boost_factor;
```

Comment:
As I mentioned in the Slack, I don't like the way this is implemented. The high pass is not implemented with any protections to ensure it's a high pass. At the end of movement initiation when the stick reaches full travel, the high pass calculation will invert and generate a large spike pulling feedforward (and the resulting pidsum) the wrong way momentarily. This could even be enough to cause a motor reversal. For example I added some debugging for just the "boost" component.

![Screen Shot 2019-07-14 at 6 47 32 PM](https://user-images.githubusercontent.com/17088539/61190312-5daded80-a668-11e9-9b30-fb2555bb34ea.png)

Top is setpoint, middle is feedforward, bottom is "boost" debug. You can clearly see the reversal caused by the lack of protection on the high-pass going negative. The bottom trace shows the "boost" contribution and you can see what it's doing to feedforward in the middle trace. If this is truly meant to be a "boost" then this needs to be fixed.

Reply:
Going negative on sudden stops is exactly the behaviour that is intended.  
When the sticks stop moving, we want the motors to slow down.  P often remains high, and I accumulates, so the only factors that slow the motors down are the D value and the drop in FF to zero.  
This PR encourages the feedforward component to drop further than it otherwise would.  The result is a stronger drop in PIDsum when the sticks stop moving than we would otherwise get.  This is useful to reduce the overshoot that we would otherwise get when the sticks stop moving, which is common, particularly at the start of a quick roll, and particularly when pidF is set relatively high. 
By adding high-pass on feed-forward, we get more early feed-forward effect, giving closer tracking, less P, and less I accumulation.  This avoids some of the overshoot tendency.  When the sticks stop, the drop in PIDsum due to feed-forward is greater than it otherwise would be, and sustained a short time.  Since the motors lag when slowing down, just like when speeding up, the negative going feedforward element ensures more of a drop in PIDsum to ensure more effective slowing down so as to avoid overshoot.
You can see the reversal in both my logs above also.  It is a good thing.
Bruce it would be nice to show set point vs gyro in your diagram above.  You will see less overshoot with this PR active than otherwise.  
It is particularly useful at the end of the flip.

Reply:
PS it is a boost in both directions.
It boosts initially in the 'positive' or 'with the stick movement' phase when the sticks are accelerating.
It also boosts in the 'negative' or 'against the stick movement' phase when the sticks are decelerating, or stopping.  
This behaviour has been shown to be useful in 'throttle boost', which has the same positive and negative boosts on accelerating and decelerating throttle.  
This behaviour helps overcome the lag motors experience both in terms of speeding up and slowing down.
I think if you fly with it on and off you'll see that both aspects are desirable.
By all means propose a different name.  I used 'boost'  in the name, since 'throttle boost' uses the same methodology so I wanted a name consistent with an existing function, and something easily and simply understood.

Reply:
Yeah I agree with @ctzsnooze that boosting in both direction is what's needed.

Reply:
Fine, then if the intent is to have it working in both directions, then the name and related settings need to be changed. It is not a "boost" in any way that makes sense to a user. Also having the "boost" and "reduction" (or whatever we want to call the end part) controlled indirectly through the same settings is not desirable. As I user it would make sense that I can feel the amount of boost on turn in and configure by adjusting the gain. But the "reduction" is accidental based on the implementation and a side effect of those same settings. If I want to use this to tune overshoot then I also have to change my turn-in feel and vice versa. I can see the argument for a "reverse push" at the end of the stick movement, but if that's part of the goal then it needs to be its own thing and not a side effect of the actual "boosting" goal.

Reply:
@etracer65 a the naming actually matches what we’ve done with throttle_boost. There the intent is similar to this case: when changing prop speed apply a transient magnification of the change in throttle to overcome the delays due to prop/motor inertia. It’s exactly the same for this Ff boost. 

So calling it boost follows what we’ve done elsewhere in the product. I personally also think you can boost an acceleration in both directions. So to me it makes sense. Kind of obvious though because I introduced throttle boost ;-).

Reply:
Just because something else is poorly named is not a good reason to perpetuate it. We need to step out of the code and the math and view this as a user would.

Reply:
Explaining to someone about how to tune overshoot caused by feedforward by increasing feedforward boost doesn't make any sense.

Reply:
@etracer65 I had several ideas for the name but none seemed any good.  FF is bigger in a log, with most moves, than without it.  So it's not wrong to think of it as a boost.  The retraction is a boost of the down-going FF component as well, but that functionality is a lot less obvious.  So while it isn't ideal, I thought boost would do.
Originally I called it FF_HPF, a technical name.  That would have absolutely zero clarity to the ordinary person and I'm glad I haven't stuck with that.
I don't think it needs much tuning.
It does allow more feedforward to be applied, generally speaking the optimal feedforward value is somewhere around 100-120.  I had thought of adding an internal multiplier within the code so that as the boost was added, the feedforward value was itself boosted.  Then the user would not need to change their current feedforward value, it would get increased 1.5x on enabling the boost.  Then it really would be a boost :-)  But in the end I thought it better to leave the two functions separate, even though to get the most out of it the user really does need to get feedforward higher than our current defaults.  Our current defaults are FF around 65.  A 1.5x boost brings that close to 100.  I currently think that 1.7 is a good start point, which brings our default FF very close to the optimal value if it was also boosted 1.7x.
If you can think of a better name, please make a suggestion. 
If you think I should boost the default feedforward value in tandem with adding the highness component, I am open to that also.

Reply:
PS one name that sort of is logical is feedforward_acceleration_factor but that's a bit tediously long.

Reply:
“Just because something else is poorly named is not a good reason to perpetuate it. We need to step out of the code and the math and view this as a user would.”

My point was rather that the name reflected the good naming scheme we had chosen in other places. I don’t agree that boost is a bad name for the concept. 

At the same time for me these are just words and I really couldn’t care less personally about how we name this stuff.

Reply:
The point is that this PR doesn't work like throttle boost. After looking at the code I'm fine with the "throttle boost" name because it's a high-pass filter on throttle and only applies a "boost" in the direction the stick is moving. But this is not what's happening with this PR. The high-pass is applied to the derivative of setpoint. So as the stick movement **slows down** a reversing factor is applied to feedforward. So there's a "boost" when the stick is accelerating, and a "reduction" when the stick is decelerating. No if this PR worked as a high-pass using setpoint (instead of the derivative of setpoint) then it would behave like throttle boost. So separate from whether this is the desired effect it shouldn't be called "feedforward boost".

Reply:
Overall I'm fine with the concept of applying a "boost" when stick is accelerating and and a "reduction" when the stick is slowing down. But as I mentioned earlier I think we need to treat these separately. The user is going to want to adjust the "boost" part for feel on initial turn-in and the "reduction" part for addressing feedforward induced overshoot - possible allowing higher overall feedforward gain. Hopefully we can setting on a fixed cutoff for the lowpass filter and simply make that a hardcoded value. Then it would make sense to have a gain for "boost" and a separate game for controlling overshoot. Exactly what this should all be called I don't know yet.

Reply:
It’s still the same thing, @etracer65: you have to realize that a constant stick
setpoint change rate merely means a constant differential in motor output on the affected axis. If you move the stick at constant positive setpoint change rate from a to b on roll the left motors have to be faster by a constant amount than the right ones. As you increase you setpoint change rate at the start of a move the motors thus have to accelerate. And at the end the have to decelerate. Ff boost has the very same effect as throttle boost here: it provides a boost shortening the amount of time required for the motors to reach the target rpm required to implement the setpoint change rate.

Reply:
And since the physical correlate of the boost is the amount of energy required to move motor rpm from a to b there is an argument to be made that it should be the same in both directions - just as it is with throttle boost. In fact we might even make this one setting between throttle boost and ff boost. That’s only right though if braking and accelerating have the same impact - meaning it takes the same boost to slow props down as to spin them up.

Reply:
And actually due to the relative nature of the pids the boost always accelerates one set of motors and decelerates the other unless airmode is required.

Reply:
@etracer65 The functionality is exactly correct, the negative-going element is needed as much as the positive-going element.  That's why the high-pass is ideal in this context.  
As to the name, I appreciate that 'boost' isn't ideal, but unless someone comes up with something better, boost it is.

Reply:
Ok, looking at the debugging of the "boost" contribution I see another concern - but possibly a way around it. In particular:

![Screen Shot 2019-07-17 at 9 22 24 AM](https://user-images.githubusercontent.com/17088539/61379031-b0f48b80-a874-11e9-90ca-6c842c1dab06.png)

The concern refers back to some of the challenges that I attempted to address with "linear feedforward". In particular the behavior of the boost shape is undesirable as it approaches the end of the stick travel as it's growing rapidly due to the derivative increasing as a result of the super-rate portion of the rates. Basically we're boosting the most at exactly the wrong time. Instead we want to be boosting early in the move and less at the end or else we're just contributing more to overshoot. Then the trace reverses as the stick slows down but it now has to fight the extra boost as well as the true feedforward component.

So possibly a way forward is to drive the "boost" off of a derivative based on the stick deflection (linear) rather than based on setpoint (including rates). In earlier testing the linear feedforward produced the more desirable shape of boosting more earlier in the move (because of expo in the rate) and less at the end (because not affected by super-rates).

And also I still think we want to be able to control the gain of the "boost" and "reduction" separately to give people a way to tune. As it is there is no way to use it to try and tame overshoot because the only way to affect that is to apply more "boost" which will contribute to more overshoot.

Reply:
@etracer65 You raise good points, which can't be answered without a long reply.

Now I have more experience flying this feature, I find that its main benefit is not primarily in stopping or reducing overshoot at the start of a flip.  Its main benefit, especially when more feedforward is applied, is in bringing gyro closer to setpoint during fast moves - reducing delay/error, especially when reversing a fast input the other way, and at the *end* more than the beginning, of a flip. Sudden, non full rate, reversals, are in particular associated with significant overshoot and setpoint to gyro delay and the quad 'runs wide'.  This code dramatically minimises that tendency.  It achieves this by allowing the use of more absolute feedforward, and adding to that early, but not late, in the move.  With double normal feedforward and a 2x boost, we get nearly four times the feedforward early in the move.  Ordinarily we cannot have that much FF early without mad overshoot.  This code provides the early benefit without worse overshoot in flips and with less overshoot and less running wide in quick direction reversals.  There are secondary benefits to I accumulation when there is less delay and error.

If we think about what this code does and doesn't do, we need to analyse the different situations where it may have a role.

First, the start of a flip or roll.
- at the very start of the input, tracking accuracy is much better, delay is less delay and error less.
- middle of the start of the the flip, strong benefits, leading to almost zero delay and flat P
- late in the start of the flip, as expo rises, the motors often can't speed up fast enough, so there is some error, but less than before.  The high-pass boost effect is easing off due to its time-constant being shorter than the input time.
- when the sticks hit max travel : the quad overshoots the same amount as before, but the overshoot is for much a shorter time.

Frankly, when the sticks hit the end of their travel, in a hard stop, this is inevitably going to overshoot for simple mechanical reasons.  The quad is rotating very, very fast, and suddenly is told to come to a dead stop.  Simple physics means such a dead stop is impossible, even with instant 100% motor reversal, since it takes time to speed up the opposing motors.  With normal inputs, now matter how hard we try to stop in an instant, our thumbs have smoothing built-in, so our the request to stop is never this hard, and good control is possible.  So the situation where the sticks hit the physical limit, abruptly,  is particularly unusual. Unless there is a predictive element that knows they are about to hit the maximum stick travel, overshoot is absolutely inevitable.  @joelucid has written code to predict getting close to max travel limit that cuts FF before reaching max travel.  This works.  However that loss of FF at high rates inevitably results in a delay in achieving the set rate at the end of the input.  Code of that kind makes for a 'pretty looking' start to a flip in a log, but adds delay at high rates.  Added delay results in greater error on the lead-up, potentially I buildup and later wobble.  

The bigger question is: does a brief overshoot at the start of a flip matter?  Should we care about it?  Assuming there isn't bounce back later, mid-flip, a very brief initial bounce-back at stick limit time cannot be seen on FPV or HD cameras, and cannot be seen visually LOS either.  What happens at the *end* of a flip does matter, but a quick overshoot at the *start* is, I think, irrelevant from the flight perspective, though it can be useful for testing and tuning purposes.  The duration of the overshoot matters more than the amount in most cases.  If the overshoot is of long duration, it can persist longer than the iterm_relax inhibition time, which leads to I accumulation once the sticks are still, and a later bounce-back mid-flip.  The iTerm windup then spoils the video when doing inverted hangs.  This code reduces that overshoot time significantly, and the negative FF often brings a compensatory reverse overshoot of equally brief duration.  This improves mid-flip or inverted hang stability by preventing unwelcome I windup. 

Therefore I think we should not be too concerned about the magnitude of the overshoot at the start of a flip, unless it is terrible; we should focus more on its duration, and avoiding delay in the lead-up; this code is beneficial in both those attributes, since the duration of the overshoot is reduced, its magnitude not made worse, and overall the risk of iTerm windup made much less. 

Let's now think about the end of a flip.  

At the start of the end of a flip, when the sticks are quickly brought back from maximum deflection, we often see significant delay in reversing the roll.  Very strong FF arises early because setpoint changes fast due to expo.  We really need this to get the motors really driving the reversal quickly. The FF boost will also be maximal due to expo.  I tested this a lot and unless the motors are already saturated, this code adds a lot of FF here and markedly reduces flip stop delay.  Because the sticks land 'softly' in the low expo region, and because the boost affect attenuates over time and is less with soft landings anyway, overshoot isn't a problem with ff_boost at the end of a flip.  Furthermore the delay reduction means less iTerm windup and less bounce back even with little iterm_relax.

The situation of fast reversals is a bit different.  Here, high FF in normal situations causes strong overshoot and some a fair bit of delay correcting the overshoot.  The quad appears to run wide or not respond as cleanly as it should.  ff_boost makes a huge difference here and does not cause overshoot.  That's because of it's self-attenuating and negative going tendency.  This provides the right FF in the right places.  It works perfectly with equal FF boost on the *outwards* direction and on the *return* direction.  There is no need at all for a separate multiplier and the complexity that such code would require.  With enough FF, the current boost implementation results in near zero delay/error, P is low and iTerm windup much less.  The tendency for the high-pass element to go negative, and the symmetrical behaviour whether going out or back, is critical and essential to the desirable behaviour.  If it didn't go negative, or was attenuated in some way on the 'return', it would not work so well.

With that background, "linear" feedforward, from stick rate rather than setpoint, has obvious drawbacks.  Currently the FF component is proportional to rate of setpoint change, and is less in the middle for small stick inputs (very small setpoint change requests) and greater for similar movements at higher stick positions (greater setpoint change requests).  In other words, the amount of FF is directly related to the requested change in rate.  This is, I think, a very good thing, especially in high-rate reversals, where attenuating FF and the boost, just because the sticks are out a bit, would make it much less useful.  And I know for sure that this is a time when we really want FF and the boost to be strong, not weak.

There is also the problem of which rate to tie a linear feedforward to.  If there is a lot of expo, and we set the linear rate to max setpoint, then there will be too much FF for small inputs in the middle.  If we tie it to the rate at centre, there won't be much FF for movements when the sticks are further out.  

I think that if FF is linear, ie determined from stick delta, not setpoint delta, it will add more problems than it solves.  I don't think it changes whether or not ff_boost is useful, since it doesn't provide the same dynamic FF benefits.  Therefore I think that providing the option of linear FF should be advanced by means of a separate PR, something unrelated to this one.

Reply:
Here are some images of sharp roll reversals.  The left panel has no ff_boost, the right does.  The quad is a high performance, high authority quad (Wuzzle's).

The first is with FF at 70.. note setpoint to delay is similar in both on and off models; the boost effect appears to do very little.

![Wuzzle_FF70_Blips](https://user-images.githubusercontent.com/11737748/61424193-0fde0180-a956-11e9-8292-793a404d59d8.jpg)

Now here is with FF at 150.  There is some overshoot, about the same in both.  But look closely at the delay.  There is far less delay with ff_boost.  In fact there is just about zero delay.  This absence of delay happens not only before the reversal, but also, most crucially, after the reversal.  After reversing direction, without boost, the quad 'runs wide' as gyro lags setpoint, and runs wide for some considerable time.  With boost, there is similar overshoot, but the negative going FF results in it being much more quickly corrected so there is no persistent 'running wide' at all.

![Wuzzle_FF150_Blips](https://user-images.githubusercontent.com/11737748/61424214-21bfa480-a956-11e9-84f9-bafa47988e95.jpg)

Here is what I'm looking at.. the integral of the delay in achieving setpoint immediately after a reversal is how much it is running wide... note how the boost effect markedly reduces that issue:

![overshoot error](https://user-images.githubusercontent.com/11737748/61424454-da85e380-a956-11e9-83f6-5c83cbfc37c9.jpg)

Reply:
From the same pilot, here is boost with FF at 200 doing rolls... note similar overshoot magnitude at the start, but look closely, that overshoot is of much shorter duration.  On the left without ff_boost the overshoot is so long as to cause iTerm windup and a slow iTerm driven correction, that would be visible in an inverted hang.  On the right, the overshoot is done and dusted much more quickly, so there is no iTerm_windup.  Further, note that there remains setpoint to gyro delay without ff_boost on the left, throughout the start of the flip and also at the end.  On the right side, with ff_boost, there is just zero delay at the start, and hardly any at the end.
![Wuzzle_FF200_Rolls2](https://user-images.githubusercontent.com/11737748/61424649-921af580-a957-11e9-895d-28343b8d5983.jpg)
This is a high performance quad with very fast reactions so it could be that the defaults are not optimal.